doc/sphinx/language/gallina-specification-language.rst

```diff
@@ -384,6 +441,22 @@ guess the missing piece of information.
 Let-in definitions
 ------------------
 
+.. insertgram term_let name_list_comma
+
+.. productionlist:: coq
+   term_let : let `name` `colon_term_opt` := `term` in `term`
+            : let `binder_list` `name` `colon_term_opt` := `term` in `term`
+            : let `single_fix` in `term`
+            : let `names_tuple` `as_return_type` := `term` in `term`
+            : let ' `pattern` := `term` `return_type_opt` in `term`
+            : let ' `pattern` in `pattern` := `term` `return_type` in `term`
```

Comment:
I hadn't realized it yet, but since `in_opt` is defined in the match section, we could further factor these two rules into a single one:

```
             : let ' `pattern` in_opt` := `term` `return_type` in `term`
```

BTW, it's amazing how inconsistent the syntax is: `let` + tuple supports a `as` and a `return` clause but the `as` clause requires the `return` clause (to be fair it is useless without the `return`), while `let` + pattern supports two independently optional clauses: the `in` and the `return` clauses. The location of the `return` varies between the two (before or after the ``` := `term` `). The two forms are syntactic sugar for `match`, which supports `as`, `in`, and `return` clauses, all independently optional.

Reply:
Ah, not quite.  One production uses `return_type_opt`, while the other uses `return_type`.

Reply:
> BTW, it's amazing how inconsistent the syntax is

May lead to an interesting discussion about whether it's beneficial to make it more consistent, particularly if that involves making some current syntax obsolete.

Reply:
Note that `let 'pat` supports the `as` clause but the latter is parsed using the syntax of pattern (since a pattern can be of the form `pat as id`).

I would suggest to open an issue to add support for the `in` clause in `let` and `if`.

I would recommend to accept the `return` clause for `let 'pat` before the `:=` so that all of `as`, `in` and `return` come together, like it is the case with `match` and `if` (and to keep the possibility of a `return` after the `:=` for compatibility).

I would lean towards accepting `as` alone (even if its impact is then limited: only telling which name to use in the synthetized return predicate).

I can do it if you wish.